File: routes/likes.js
const express = require('express');

module.exports = (db) => {
  const router = express.Router();

  router.post('/like', (req, res) => {
    const { fromUser, toUser } = req.body;
    if (!fromUser || !toUser) return res.status(400).json({ success: false, error: 'fromUser и toUser обязательны' });
    db.get("SELECT likes, matches FROM users WHERE userId = ?", [fromUser], (err, row) => {
      if (err || !row) return res.status(404).json({ success: false, error: 'Отправитель не найден' });
      let likes = [];
      let matches = [];
      try { likes = JSON.parse(row.likes || '[]'); } catch (e) {}
      try { matches = JSON.parse(row.matches || '[]'); } catch (e) {}
      if (!likes.includes(toUser)) {
        likes.push(toUser);
        db.run("UPDATE users SET likes = ? WHERE userId = ?", [JSON.stringify(likes), fromUser]);
      }
      db.get("SELECT likes, matches FROM users WHERE userId = ?", [toUser], (err2, row2) => {
        if (err2 || !row2) return res.status(404).json({ success: false, error: 'Получатель не найден' });
        let likes2 = [];
        let matches2 = [];
        try { likes2 = JSON.parse(row2.likes || '[]'); } catch (e) {}
        try { matches2 = JSON.parse(row2.matches || '[]'); } catch (e) {}
        let isMatch = false;
        if (likes2.includes(fromUser)) {
          isMatch = true;
          if (!matches.includes(toUser)) {
            matches.push(toUser);
            db.run("UPDATE users SET matches = ? WHERE userId = ?", [JSON.stringify(matches), fromUser]);
          }
          if (!matches2.includes(fromUser)) {
            matches2.push(fromUser);
            db.run("UPDATE users SET matches = ? WHERE userId = ?", [JSON.stringify(matches2), toUser]);
          }
        }
        res.json({ success: true, match: isMatch });
      });
    });
  });

  router.post('/dislike', (req, res) => {
    const { fromUser, toUser } = req.body;
    if (!fromUser || !toUser) return res.status(400).json({ success: false, error: 'fromUser и toUser обязательны' });
    db.get("SELECT dislikes FROM users WHERE userId = ?", [fromUser], (err, row) => {
      if (err || !row) return res.status(404).json({ success: false, error: 'Отправитель не найден' });
      let dislikes = [];
      try { dislikes = JSON.parse(row.dislikes || '[]'); } catch (e) {}
      if (!dislikes.includes(toUser)) {
        dislikes.push(toUser);
        db.run("UPDATE users SET dislikes = ? WHERE userId = ?", [JSON.stringify(dislikes), fromUser], (err2) => {
          if (err2) return res.status(500).json({ success: false, error: err2.message });
          res.json({ success: true });
        });
      } else {
        res.json({ success: true });
      }
    });
  });

  router.post('/superlike', (req, res) => {
    const { senderId, receiverId } = req.body;
    console.log(`[routes/likes.js] POST /superlike: senderId ${senderId} to receiverId ${receiverId}`);

    if (!senderId || !receiverId) {
      return res.status(400).json({ success: false, error: 'senderId и receiverId обязательны' });
    }

    // Проверяем, что отправитель существует
    db.get("SELECT superlikes FROM users WHERE userId = ?", [senderId], (err, senderRow) => {
      if (err || !senderRow) {
        console.error('❌ Ошибка в /superlike:', { message: err && err.message, stack: err && err.stack });
        return res.status(404).json({ success: false, error: 'Отправитель не найден' });
      }
      let superlikes = [];
      try { superlikes = JSON.parse(senderRow.superlikes || '[]'); } catch (e) {}
      if (superlikes.includes(receiverId)) {
        return res.json({ success: true, message: 'Суперлайк уже существует' });
      }
      superlikes.push(receiverId);
      db.run("UPDATE users SET superlikes = ? WHERE userId = ?", [JSON.stringify(superlikes), senderId], (err2) => {
        if (err2) {
          console.error('❌ Ошибка при обновлении superlikes:', err2.message);
          return res.status(500).json({ success: false, error: err2.message });
        }
        // Можно добавить дополнительную логику для мэтча по суперлайку, если нужно
        res.json({ success: true });
      });
    });
  });

  router.get('/likesReceived', (req, res) => {
    const { userId } = req.query;
    console.log(`[routes/likes.js] GET /likesReceived: userId='${userId}'`);

    if (!userId) {
      return res.status(400).json({ success: false, error: 'userId обязателен' });
    }

    const sql = `
      SELECT l.*, u.username, u.photo1, u.photo2, u.photo3, u.bio, u.age, u.location
      FROM likes l
      JOIN users u ON l.fromUser = u.userId
      WHERE l.toUser = ?
      ORDER BY l.timestamp DESC
    `;

    db.all(sql, [userId], (err, rows) => {
      if (err) {
        console.error(`[routes/likes.js] GET /likesReceived error: ${err.message}`);
        return res.status(500).json({ success: false, error: err.message });
    }

      res.json({ success: true, likes: rows });
    });
  });

  router.get('/likesMade', (req, res) => {
    const { userId } = req.query;
    console.log(`[routes/likes.js] GET /likesMade: userId='${userId}'`);

    if (!userId) {
      return res.status(400).json({ success: false, error: 'userId обязателен' });
    }

    const sql = `
      SELECT l.*, u.username, u.photo1, u.photo2, u.photo3, u.bio, u.age, u.location
      FROM likes l
      JOIN users u ON l.toUser = u.userId
      WHERE l.fromUser = ?
      ORDER BY l.timestamp DESC
    `;

    db.all(sql, [userId], (err, rows) => {
      if (err) {
        console.error(`[routes/likes.js] GET /likesMade error: ${err.message}`);
        return res.status(500).json({ success: false, error: err.message });
    }

      res.json({ success: true, likes: rows });
    });
  });

  return router;
};

File: python_backend/routes/likes.py
"""
routes/likes.py
Роуты для лайков, дизлайков и суперлайков
"""
from fastapi import APIRouter, Query, HTTPException, Body
from typing import Dict, List
from pydantic import BaseModel
from db_utils import db_get, db_all, db_run, safe_json_parse
import json

router = APIRouter()


class LikeRequest(BaseModel):
    fromUser: str
    toUser: str


class DislikeRequest(BaseModel):
    fromUser: str
    toUser: str


class SuperlikeRequest(BaseModel):
    senderId: str
    receiverId: str


@router.post("/like")
async def like_user(data: LikeRequest):
    """Поставить лайк пользователю"""
    if not data.toUser:
        raise HTTPException(status_code=400, detail="toUser обязателен")
    
    # Получаем данные отправителя
    from_user_row = await db_get("SELECT likes, matches FROM users WHERE userId = ?", [data.fromUser])
    if not from_user_row:
        raise HTTPException(status_code=404, detail="Отправитель не найден")
    
    likes = safe_json_parse(from_user_row.get("likes"), [])
    matches = safe_json_parse(from_user_row.get("matches"), [])
    
    # Добавляем лайк, если его еще нет
    if data.toUser not in likes:
        likes.append(data.toUser)
        await db_run("UPDATE users SET likes = ? WHERE userId = ?", [json.dumps(likes), data.fromUser])
    
    # Проверяем взаимный лайк
    to_user_row = await db_get("SELECT likes, matches FROM users WHERE userId = ?", [data.toUser])
    if not to_user_row:
        raise HTTPException(status_code=404, detail="Получатель не найден")
    
    likes2 = safe_json_parse(to_user_row.get("likes"), [])
    matches2 = safe_json_parse(to_user_row.get("matches"), [])
    is_match = False
    
    if data.fromUser in likes2:
        is_match = True
        if data.toUser not in matches:
            matches.append(data.toUser)
            await db_run("UPDATE users SET matches = ? WHERE userId = ?", [json.dumps(matches), data.fromUser])
        if data.fromUser not in matches2:
            matches2.append(data.fromUser)
            await db_run("UPDATE users SET matches = ? WHERE userId = ?", [json.dumps(matches2), data.toUser])
    
    return {"success": True, "match": is_match}


@router.post("/dislike")
async def dislike_user(data: DislikeRequest):
    """Поставить дизлайк пользователю"""
    if not data.toUser:
        raise HTTPException(status_code=400, detail="toUser обязателен")
    
    from_user_row = await db_get("SELECT dislikes FROM users WHERE userId = ?", [data.fromUser])
    if not from_user_row:
        raise HTTPException(status_code=404, detail="Отправитель не найден")
    
    dislikes = safe_json_parse(from_user_row.get("dislikes"), [])
    if data.toUser not in dislikes:
        dislikes.append(data.toUser)
        await db_run("UPDATE users SET dislikes = ? WHERE userId = ?", [json.dumps(dislikes), data.fromUser])
    
    return {"success": True}


@router.post("/superlike")
async def superlike_user(data: SuperlikeRequest):
    """Поставить суперлайк пользователю"""
    if not data.receiverId:
        raise HTTPException(status_code=400, detail="receiverId обязателен")
    
    # В БД колонка называется superLikesCount (camelCase)!
    sender_row = await db_get("SELECT \"superLikesCount\" FROM users WHERE \"userId\" = ?", [data.senderId])
    if not sender_row:
        raise HTTPException(status_code=404, detail="Отправитель не найден")
    
    # TODO: Реализовать логику суперлайков (проверка лимита, запись в БД)
    
    return {"success": True}


@router.get("/likesReceived")
async def get_likes_received(userId: str = Query(..., description="ID пользователя")):
    """Получить количество пользователей, которые лайкнули текущего пользователя"""
    # Находим всех пользователей, у которых в likes есть userId текущего пользователя
    # Используем jsonb_array_elements_text для развертывания массива и проверки наличия значения
    try:
        sql = """
            SELECT COUNT(*) as count
            FROM users
            WHERE EXISTS (
                SELECT 1
                FROM jsonb_array_elements_text(likes::jsonb) AS elem
                WHERE elem = %s
            )
        """
        result = await db_get(sql, [userId])
        
        if not result:
            return {"success": True, "count": 0}
        
        count = result.get("count", 0)
        return {"success": True, "count": count}
    except Exception as e:
        print(f"[GET /api/likesReceived] Ошибка: {e}")
        import traceback
        traceback.print_exc()
        return {"success": True, "count": 0}


@router.get("/likesMade")
async def get_likes_made(userId: str = Query(..., description="ID пользователя")):
    """Получить список пользователей, которым поставил лайк текущий пользователь"""
    user_row = await db_get("SELECT likes FROM users WHERE userId = ?", [userId])
    if not user_row:
        return {"success": True, "likes": []}
    
    liked_user_ids = safe_json_parse(user_row.get("likes"), [])
    if not liked_user_ids:
        return {"success": True, "likes": []}
    
    placeholders = ",".join(["?"] * len(liked_user_ids))
    sql = f"""
        SELECT userId, username, photo1, photo2, photo3, bio, age, name
        FROM users
        WHERE userId IN ({placeholders})
    """
    rows = await db_all(sql, liked_user_ids)
    
    return {"success": True, "likes": rows}

